### ISSUE TITLE
[Inductor] Remove Redundant NaN Propagation in `max` op in Softmax

### ISSUE THREAD

**Issue by liqiangxl** [pytorch/pytorch#189161]
### 🚀 The feature, motivation and pitch

# Redundant NaN Propagation in `max` op in non-onlined Softmax

## Issue

Current Inductor softmax can do redundant NaN propagation in the softmax-internal max reduction.

For the fallback/persistent softmax form, the computation is effectively:

```python
xmax = max(x, dim=-1, keepdim=True)
xexp = exp(x - xmax)
xsum = sum(xexp, dim=-1, keepdim=True)
y = xexp / xsum
```

The internal `max` is only used to stabilize softmax. It is not returned to the user as a standalone `amax` result.

Today, that max can lower to NaN-propagating max operations such as:

```python
tl.maximum(a, b, tl.PropagateNan.ALL)
```

This is redundant for the fallback/persistent softmax path. If an input row contains a NaN, the original NaN input is still consumed by `x - xmax`. That NaN flows through `exp`, poisons `xsum`, and makes the final softmax row NaN. The final softmax behavior is preserved even if the internal max ignores NaNs.

So the softmax-local max can use `fmax` semantics:

```python
xmax = fmax(x, dim=-1, keepdim=True)
xexp = exp(x - xmax)
xsum = sum(xexp, dim=-1, keepdim=True)
y = xexp / xsum
```

and lower to cheaper operations like:

```python
tl.maximum(a, b)
triton_helpers.fmax2(...)
```

## Performance Impact

The redundant NaN-propagating max has measurable cost. On GB200, with:

```python
x = torch.randn(32768, N, device="cuda", dtype=torch.bfloat16)
```

and online softmax enabled, start with `N=1024`. This shape already uses the persistent reduction path by default, so the effect of removing redundant NaN propagation is visible directly:

```text
main/default                  27.10 us
softmax-nan/default           25.82 us   1.05x speedup
```

For larger `N`, the default lowering does not select the persistent kernel, so `main` and `softmax-nan` are mostly identical unless `TORCHINDUCTOR_MULTI_KERNEL=1` is used. That knob exposes/selects the persistent multi-kernel candidate. Once both branches are on that persistent path, the speedup from removing redundant NaN propagation becomes visible.

At `N=2048`, the measured progression was:

```text
main/default                  111.6 us
main + multi-kernel             75.59 us
softmax-nan + multi-kernel      64.29 us
```

This separates two effects:

- `TORCHINDUCTOR_MULTI_KERNEL=1` selects the faster persistent path: `111.6 -> 75.59 us`.
- Removing redundant NaN propagation from the softmax-local max further improves that persistent path: `75.59 -> 64.29 us`.

Across the sweep, `llu/softmax-nan + TORCHINDUCTOR_MULTI_KERNEL=1` was `1.15x` geomean faster than `main + TORCHINDUCTOR_MULTI_KERNEL=1`.

### Performance Plot

<img width="1003" height="533" alt="Image" src="https://github.com/user-attachments/assets/2c662351-99f6-4c4a-be2c-317fc301dce8" />

The figure plots SOL% using the bf16 softmax lower-bound traffic, `4 * M * N` bytes for one input read plus one output write, divided by an assumed GB200 per-GPU HBM peak of `8.0 TB/s`. The gray line is `main` with default `TORCHINDUCTOR_MULTI_KERNEL=0`, the blue line is `main` with `TORCHINDUCTOR_MULTI_KERNEL=1`, and the red line is `llu/softmax-nan` with `TORCHINDUCTOR_MULTI_KERNEL=1`.

### Performance Table

| N | main default us | softmax-nan default us | main MK=1 us | softmax-nan MK=1 us | softmax-nan/default | softmax-nan/MK=1 |
|---:|---:|---:|---:|---:|---:|---:|
| 1024 | 27.10 | 25.82 | 27.10 | 25.79 | 1.05x | 1.05x |
| 2048 | 111.64 | 111.65 | 75.59 | 64.29 | 1.00x | 1.18x |
| 3072 | 186.02 | 185.98 | 118.15 | 97.76 | 1.00x | 1.21x |
| 4096 | 218.47 | 218.62 | 120.97 | 107.91 | 1.00x | 1.12x |
| 5120 | 286.00 | 286.07 | 162.38 | 140.03 | 1.00x | 1.16x |
| 6144 | 333.51 | 333.56 | 176.48 | 161.44 | 1.00x | 1.09x |
| 7168 | 380.87 | 380.80 | 214.14 | 178.88 | 1.00x | 1.20x |
| 8192 | 409.19 | 409.16 | 235.23 | 194.76 | 1.00x | 1.21x |
| 9216 | 475.87 | 475.47 | 341.47 | 289.69 | 1.00x | 1.18x |
| 10240 | 523.12 | 523.01 | 344.86 | 297.07 | 1.00x | 1.16x |
| 11264 | 570.81 | 570.52 | 349.30 | 303.80 | 1.00x | 1.15x |
| 12288 | 617.20 | 618.14 | 390.15 | 390.85 | 1.00x | 1.00x |
| 13312 | 664.78 | 665.14 | 453.66 | 408.93 | 1.00x | 1.11x |
| 14336 | 711.62 | 714.11 | 474.47 | 417.89 | 1.00x | 1.14x |
| 15360 | 760.19 | 761.28 | 512.17 | 426.27 | 1.00x | 1.20x |
| 16384 | 767.55 | 767.98 | 556.45 | 429.19 | 1.00x | 1.30x |

## What Must Not Change

This optimization should not be applied globally.

NaN propagation is still required for:

- generic `aten.amax`, because its result is user-visible
- generic max reductions whose result may be observed directly
- online softmax max-state reductions, where the algorithm combines partial `(max, sum)` states and can otherwise lose the fact that a row contained a NaN

The intended scope is narrow: use non-NaN-propagating `fmax` only for the softmax-local fallback/persistent max reduction where the original input `x` still flows into the subsequent `exp(x - xmax)` and `sum` computation.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by liqiangxl** [comment]
Since `TORCHINDUCTOR_MULTI_KERNEL=1` leads to geo-mean speedup of 1.55x, we may consider auto enable it when compiled with max-autotune.

**Comment by kshitij12345** [comment]
Can this be closed given that https://github.com/pytorch/pytorch/issues/189162 was merged?

**Comment by kshitij12345** [comment]
> Since TORCHINDUCTOR_MULTI_KERNEL=1 leads to geo-mean speedup of 1.55x, we may consider auto enable it when compiled with max-autotune.

Makes sense to me , however I think this should be a separate issue.
